### PR TITLE
Implement application-wide state and update flow. Unittest <DateInput>.

### PR DESCRIPTION
**Pull request by awong-dev**
Proof of concept for storing application wide state with a subcomponent.

Also has basic validation code to DateInput and unittests to verify the component behaves properly.

### CODE REVIEW

File: _health-care/_js/_components/_health-care-app.jsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        day: "5",
+        year: "2014",
+      };
+    }
+  }
+
+  onStateChange(field, update) {
+    let newApplicationData = Object.assign({}, this.state.applicationData);
+    newApplicationData[field] = update;
+    window.localStorage["health-app-data"] = JSON.stringify(newApplicationData);
```
Comment (akainic) [inline _health-care/_js/_components/_health-care-app.jsx]: Can you not reference `savedAppData` here instead?

File: _health-care/_js/_components/_health-care-app.jsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        <Nav/>
+        {
+          // Pass the application state down to child components. The
+          // cloneElement idiom is the standard react way to add
+          // properties to a child. It doesn't actually new a bunch of objects
+          // so it's fast.
+          React.Children.map(this.props.children, (child) => {
+              return React.cloneElement(child,
+                { applicationData: this.state.applicationData,
+                  onStateChange: this.onStateChange })
```
Comment (akainic) [inline _health-care/_js/_components/_health-care-app.jsx]: So this is creating props for all of `HealthCareApp`'s children, which may not reach the question components. Will we have to add this to every layer so it reaches the bottom-most components?

File: _health-care/_js/_components/_hello-world.jsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+class Hello extends React.Component {
+  render() {
+    return(
+      <div>
+        <h2>Hello World</h2>
+
+        <div className="row">
+          <h3>Happy Date</h3>
+          <DateInput date={this.props.applicationData.happyDate}
```
Comment (akainic) [inline _health-care/_js/_components/_hello-world.jsx]: Eventually, when `Hello` is actually a panel or block of questions (some intermediary component), should it have access to all of `applicationData`? Is there a way for each layer to only know about the data for the layers it contains? Maybe `applicationData` has nested objects within it that correspond to different layers? This is just thinking in advance

File: _health-care/_js/_components/_health-care-app.jsx (hunk 1 above)
Comment (awong-dev) [inline _health-care/_js/_components/_health-care-app.jsx]: In react, you're supposed to consider `this.state` to be immutable and only update it via `this.setState()`. Thus, I made a copy.

File: _health-care/_js/_components/_health-care-app.jsx (hunk 2 above)
Comment (awong-dev) [inline _health-care/_js/_components/_health-care-app.jsx]: Correctish.

I assume each Page will more explicitly pass a subset of structure down to their Question component which will then filter down into the input fields.

It doesn't all quite work yet and is clunky, but this is indeed how data is supposed to propagate.

Another option is to use `window.applicationData`, but that's a global variable and I hate globals.  (Yes...applicationData is also a global variable effectively, but the `setState()` + `props` patterns restrict mutation and inspection so it's actually easier to reason about still.)

File: _health-care/_js/_components/_hello-world.jsx (hunk 3 above)
Comment (awong-dev) [inline _health-care/_js/_components/_hello-world.jsx]: Good questions...I'm struggling with this too.

I'm thinking of Hello as a panel. I think each panel should have full access to applicationData and then pass a sub-struct down to the Question. So you could imagine applicationData.personalInfo.{birthDate,name, etc} and inside the PersonalInformation panel,you'd pass `applicationData.personalInfo.birthDate` to the BirthDate question which gives you the information hiding.

Comment (awong-dev) [inline _health-care/_js/_components/_hello-world.jsx]: But there will be a problem on the state update. :-/  I was going to try to work through that today.

I know how to fix it using layers of functions...it just feels a little ugly.

File: _health-care/_js/_components/date_input.jsx [hunk 4]
```diff
@@ -4,54 +4,72 @@ import _ from 'lodash';
...
-  }
+  handleChange() {
+    const date = {
+        month: this.refs.month.value,
+        day: this.refs.day.value,
+        year: this.refs.year.value
+      }
+
+    // Update UI based on validation.
+    if (!this.validate(date.day, date.month, date.year)) {
```
Comment (akainic) [inline _health-care/_js/_components/date_input.jsx]: Wonder if we should also call `validate` on load or something, I noticed that if the page is reloaded and the input field is empty it won't be marked as having an error.

Comment (awong-dev) [inline _health-care/_js/_components/date_input.jsx]: You're probably right. Let me do in a followup PR. This one's too big already.

Comment (akainic) [inline _health-care/_js/_components/date_input.jsx]: Sounds good

**Comment by akainic** [comment]
LGTM!

**Pull request merged**